Make homepage content translatable

mlovic opened the pull request:
__NOTE:__ For reasons explained in meta-issue https://github.com/AyuntamientoMadrid/consul/issues/1612, this branch contains commits which are already included in other open PRs. After those PRs are merged, only the commits _after_ 80b4b3e will be left. Those are the commits which need to be reviewed.

__Do not merge before reading https://github.com/AyuntamientoMadrid/consul/issues/1612__

References
===================
 - Original issue: https://github.com/consul/consul/issues/2668
 - PR used as reference: https://github.com/consul/consul/pull/2619

Objectives
===================
Allow `Widgets::Card`s to exist in multiple languages. This applies to 4 of its fields: label, title, description, link_text. See commits:

76ba3c0 Make Widget::Card translatable
27f2fdb Remove extra tag element from cards form
59c2cbe  Add Spanish translation to Card seed data

Visual Changes
===================
Same as all other admin "translatable" forms

![peek 2018-07-30 15-46](https://user-images.githubusercontent.com/7111622/43401237-e34fa53a-940f-11e8-8e5c-9d70053a97fd.gif)

This PR also adds Spanish translations to the seed Cards. This will be visible after generating the dev seed data when using the application in Spanish (the default).

**Before**

![selection_121](https://user-images.githubusercontent.com/7111622/44837302-8b00d100-ac39-11e8-9ca7-de720588c865.png)

**After**

![selection_122](https://user-images.githubusercontent.com/7111622/44837311-918f4880-ac39-11e8-846a-e4f069c7be58.png)

Notes
===================
As with the original [Milestones PR](https://github.com/consul/consul/pull/2619), data migration is not implemented: 

> At this moment, the migration that creates the milestones translations table is not migrating the existing data. If this is the desired behaviour, it's OK with it. If not, the migration should have the :migrate_data => true option, as seen in the globalize readme:

~This PR also builds on changes from  https://github.com/AyuntamientoMadrid/consul/pull/1585 which has no been merged yet. I cherry-picked the relevant commit here.~ __Edit:__: those changes are already in master (701c1e9)

javierm commented on app/models/widget/card.rb:
```diff
@@ -1,8 +1,15 @@
+  translates :title,       touch: true
+  translates :description, touch: true
+  translates :link_text,   touch: true
+  globalize_accessors locales: [:en, :es, :fr, :nl, :val, :pt_br]
```
Hi, @mlovic. Since the languages seem to always be `[:en, :es, :fr, :nl, :val, :pt_br]` on this project and `[:en, :es, :fr, :nl, :pt_br]` in CONSUL's main repo, I was wondering if it makes sense to have them in a configuration file. That way we remove duplication, and backporting these PRs will be easier. What do you think?

mlovic replied:
Good idea. It looks like those are the same locales that are [registered with I18n](https://github.com/AyuntamientoMadrid/consul/blob/master/config/application.rb#L23), except that 'pt-BR' needs to be underscored ([apparently](https://github.com/AyuntamientoMadrid/consul/pull/1435) this gem does not handle uppercase or `-`).

How about:

```ruby
  globalize_accessors locales: I18n.available_locales.map(&:to_s).map(&:underscore)
```

javierm replied:
I hadn't realized they were already defined. Great catch!

aitbw replied:
Re: `pt-BR`, @iagirre documented this case very well on consul#2619

mlovic replied:
Thanks @aitbw :+1: . I tried to link to that from my last message.

javierm commented:
@mlovic I think we usually rebase against master instead of adding a merge commit. Correct me if I'm wrong!

mlovic commented on app/controllers/admin/widget/cards_controller.rb:
```diff
@@ -41,11 +42,20 @@ def destroy
-end
\ No newline at end of file
+  def resource
+   Widget::Card.find(params[:id])
```
Good catch!

mlovic commented:
> @mlovic I think we usually rebase against master instead of adding a merge commit. Correct me if I'm wrong!

Actually I don't know, but you're probably right. I'm happy to change it.

javierm commented on app/views/admin/widget/cards/_form.html.erb:
```diff
@@ -1,14 +1,58 @@
+                         class: "js-globalize-attribute",
+                         data: {locale: locale},
+                         style: display_translation?(locale),
+                         label: false %>
```
A couple of suggestions regarding the text fields.

First, instead of using a generic label which isn't linked to a text field, what about removing the `f.label` lines and using something like:
```ruby
 <%= f.text_field "title_#{locale}",
                  placeholder: t("admin.banners.banner.title"),
                  class: "js-globalize-attribute",
                  data: {js_banner_title: "js_banner_title",
                         locale: locale},
                  style: display_translation?(locale),
                  label: t("admin.banners.banner.title"),
                  label_options: {
                    style: display_translation?(locale),
                    class: "js-globalize-attribute",
                    data: { locale: locale }
                   } %>
```
This way we can keep the benefits of using HTML labels (clicking on the label takes you to the field, ...).

Second, how about extracting this logic into a method? Maybe something like `f.translated_text field`?

mlovic replied:
> This way we can keep the benefits of using HTML labels (clicking on the label takes you to the field, ...).

I wasn't aware of this behavior. Sure, I'll see if it works with the current JS implementation and change it.

> Second, how about extracting this logic into a method? Maybe something like f.translated_text field?

I thought about this a while ago, and started coding something. The tricky part is merging the relevant options. Here is what I was thinking:

```ruby
module TranslatableFormHelper
  def translatable_form_for(record, options = {}, &block)
    form_for(record, options.merge(builder: TranslatableFormBuilder), &block)
  end
end

class TranslatableFormBuilder < ActionView::Helpers::FormBuilder
  def translatable_text_field(method, _options = {})
    options = _options.dup
    options[:class] = (options[:class] || "") + " js-globalize-attribute" 
    options[:data] = (options[:data] || {}).merge(locale: locale)
    options[:label] = false
    options[:style] = display_translation?(locale)

    @template.capture do
      @object.globalize_locales.each do |locale|
        globalize(locale) do
          @template.concat do
            text_field(method, options.merge(title: _options[:title] + "_" + locale))
          end
        end
      end
    end

    text_field(method)
  end
```

javierm replied:
I think we're using a custom form builder (check `config/initializers/foundation_rails_helper.rb`). Maybe we could add it there, next to `cktext_area`?

I'd keep the `each` loop and the `capture` and `globalize` part in another method (maybe even outside the builder? I'm not sure), but since it's a matter of personal preference, I'll leave it up to you.

mlovic replied:
Well another reason to use a custom `FormBuilder` is to handle the "deleted_<locale>" tags that every translatable form needs.

I'll consider your other suggestion, code something up quickly and see what you think.

javierm replied:
I think you're right! Just one thing; instead of:

`class TranslatableFormBuilder < ActionView::Helpers::FormBuilder`

I think it would be safer to inherit from:

`class TranslatableFormBuilder < FoundationHelpers::FormBuilder`

Since that's the builder we use in the rest of the application (or so I think).

mlovic replied:
Here is a working implementation of the FormBuilder solution: https://github.com/AyuntamientoMadrid/consul/commit/7e38dc7c0caf5d8daa29c43aa8c96790c421b58b. What do you think @javierm? 

I'll have to open a new PR to apply these changes to all the translatable forms in admin. @javierm do you want to merge this branch as it is, and wait for the dedicated refactor PR to make these changes? Or do you want me to add the changes here (and then also open a refactor PR to make the changes to other forms)?

javierm replied:
@mlovic Looks great! :tada:. One small thing: would it be possible to remove the `f.label` calls and include something like what's suggested in https://github.com/AyuntamientoMadrid/consul/pull/1588#discussion_r208894203 in the `translatable_field` method? I tested the `label_options` locally and they seemed to work fine with globalize, but maybe I missed something.

Regarding PRs, I'll leave a comment in #1602.

javierm replied:
Regarding where to add the translatable form builder, IMHO you could either add it to #1602 or open a new one which adds the builder and applies it to the already merged banners. Up to you!

mlovic replied:
Several of these suggestions have been implemented in https://github.com/AyuntamientoMadrid/consul/pull/1613.

javierm commented:
> >@mlovic I think we usually rebase against master instead of adding a merge commit. Correct me if I'm wrong!
>
> Actually I don't know, but you're probably right. I'm happy to change it.

@aitbw I think you know what we usually do with merge/rebase. What do you think?

aitbw commented:
+1 to rebase over merging `master`

aitbw commented:
I have just one question: You said on the PR's description that you cherry picked a commit from #1585 so you can start building from there because that commit is on a PR that hasn't been merged yet but it's indeed merged and I don't see a commit from that pull request that could be backported here :thinking:

Could you shed some light on this, @mlovic ?

mlovic commented:
@aitbw You're right, I can't remember if I rebased that commit out or what...

In any case, the commit in question is already in master (https://github.com/AyuntamientoMadrid/consul/pull/1585/commits/701c1e94aae51fb39ffd84a9022c70f3b2e8bf62, different SHA because I rebased before merging), so I'll just remove that note from the description.

mlovic commented:
Rebasing this branch onto the "translatable" refactor branch . Note that all commits except the last one should be ignored as they are already included in other PRs. See https://github.com/AyuntamientoMadrid/consul/issues/1612 for explanation and for merge instructions.

javierm commented on app/views/admin/widget/cards/_form.html.erb:
```diff
@@ -1,14 +1,18 @@
   </div>
 
-  <%= f.text_field :title %>
+  <%= f.translatable_text_field :title %>
```
I think we either need to add the label option like in banners, where we add `label: t("admin.banners.banner.title")`, or change something in translatable form in order to find the correct label automatically.

mlovic replied:
Right! I did this a while ago by adding `label: @object.class.human_attribute_name(method),` to the form builder method, but I later removed it as I couldn't remember why it was necessary (it's so that the label is rendered as "Title" instead of "Title en"). 

I'll add it back.

mlovic replied:
Fixed by 49b9ae7

javierm commented on app/views/admin/widget/cards/_form.html.erb:
```diff
@@ -1,14 +1,18 @@
   <div class="small-12 medium-6">
-    <%= f.text_field :link_text %>
+    <%= f.translatable_text_field :link_text %>
+    <%= f.label :link_text %>
```
I think there's a typo and we should remove this line.

javierm commented on spec/features/admin/widgets/cards_spec.rb:
```diff
@@ -135,4 +135,115 @@ def attach_image_to_card
-end
\ No newline at end of file
+
+  context "Translations" do
```
I've seen we've got specs using `it_behaves_like` and `shared_examples` methods. Do you think it makes sense to do something similar with `translatable` specs?

Sorry I didn't comment about it earlier, since it could have been included in #1613; I only saw the duplication now that I was reviewing this PR.

mlovic replied:
Yes, there is quite a bit of test duplication across resources :/. @voodoorai2000 said to leave it that way for now and worry about refactoring to shared examples later. I might try to do it anyway though if I have time.

javierm commented:
> As with the original Milestones PR, data migration is not implemented

@mlovic Very reasonable. Do you happen to know why it wasn't implemented originally and how we're going to deal with existing data? If you don't know, sorry I'm asking! Rai and Itziar aren't around, so...

mlovic commented:
> @mlovic Very reasonable. Do you happen to know why it wasn't implemented originally and how we're going to deal with existing data? If you don't know, sorry I'm asking! Rai and Itziar aren't around, so...

@javierm I don't know. I did the same for the (merged) [Banners PR](#1585) and Rai didn't comment on it. I'll make a note on #1612.

MariaCheca commented:
This PR is being closed in favor of https://github.com/AyuntamientoMadrid/consul/pull/1633, which includes the last 3 commits and solves schema and specs conflicts.
